presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+    public long sizeOf()
+    {
+        return INSTANCE_SIZE
+                + groupIdToHeapBuffer.sizeOf()
+                + heapNodeBuffer.sizeOf()
+                + peerGroupBuffer.sizeOf()
+                + heapTraversal.sizeOf()
+                + peerGroupLookup.sizeOf();
+    }
+
+    private long calculateRootRank(long groupId)
+    {
+        long heapValueCount = groupIdToHeapBuffer.getHeapValueCount(groupId);
+        long heapRootIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        checkArgument(heapRootIndex != UNKNOWN_INDEX, "Group does not have a root");
+        long rootPeerGroupCount = heapNodeBuffer.getPeerGroupCount(heapRootIndex);
+        return heapValueCount - rootPeerGroupCount + 1;
+    }
+
+    /**
+     * Add the specified row to this accumulator.
+     * <p>
+     * This may trigger row eviction callbacks if other rows have to be evicted to make space.
+     *
+     * @return true if this row was incorporated, false otherwise
+     */
+    public boolean add(long groupId, long newRowId)
+    {
+        checkArgument(newRowId != nullRowId, "Cannot add null row ID");
+
+        // Insert to any existing peer groups first (heap nodes contain distinct values)
+        long peerHeapNodeIndex = peerGroupLookup.get(groupId, newRowId);
+        if (peerHeapNodeIndex != UNKNOWN_INDEX) {
+            directPeerGroupInsert(groupId, peerHeapNodeIndex, newRowId);
+            if (calculateRootRank(groupId) > topN) {
+                heapPop(groupId, rowIdEvictionListener);
+            }
+            // Return true because heapPop is guaranteed not to evict the newly inserted row (by definition of rank)
+            return true;
+        }
+
+        groupIdToHeapBuffer.allocateGroupIfNeeded(groupId);
+        if (groupIdToHeapBuffer.getHeapValueCount(groupId) < topN) {
+            // Always safe to insert if total number of values is still less than topN
+            long newPeerGroupIndex = peerGroupBuffer.allocateNewNode(newRowId, UNKNOWN_INDEX);
+            heapInsert(groupId, newPeerGroupIndex, 1);
+            return true;
+        }
+        else if (rowComparisonStrategy.compare(newRowId, peekRootRowId(groupId)) < 0) {
+            // Given that total number of values >= topN, we can only consider values that are less than the root (otherwise topN would be violated)
+            long newPeerGroupIndex = peerGroupBuffer.allocateNewNode(newRowId, UNKNOWN_INDEX);
+            // Rank will increase by +1 after insertion, so only need to pop if root rank is already == topN.
+            if (calculateRootRank(groupId) < topN) {
+                heapInsert(groupId, newPeerGroupIndex, 1);
+            }
+            else {
+                heapPopAndInsert(groupId, newPeerGroupIndex, 1, rowIdEvictionListener);
+            }
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    private void directPeerGroupInsert(long groupId, long heapNodeIndex, long rowId)
```

Comment:
move private methods below public

Reply:
This one is used directly by the method above, so in that sense it makes sense to retain that locality here